### ISSUE TITLE
Doxygen runs infinitely in multi-thread build

### ISSUE THREAD

**Issue by arwedus** [doxygen/doxygen#9151]
**Describe the bug**

With multithreading enabled, doxygen runs into an infinite loop in the "patching output file" stage.

See for example a CI console output log:

```
...
2022-02-23T10:36:09.4091787Z Patching output file 1435/1440
2022-02-23T10:36:09.4092448Z Patching output file 1436/1440
2022-02-23T10:36:09.4093095Z Patching output file 1437/1440
2022-02-23T10:36:09.4094036Z Patching output file 1438/1440
2022-02-23T10:36:09.4094671Z Patching output file 1439/1440
2022-02-23T10:36:09.4095326Z Patching output file 1440/1440
2022-02-23T10:36:09.4095986Z Patching output file 1441/1440
2022-02-23T10:36:09.4096644Z Patching output file 1442/1440
2022-02-23T10:36:09.4097296Z Patching output file 1443/1440
2022-02-23T10:36:09.4097980Z Patching output file 1444/1440
...
```

this continues counting up until I cancelled it.

**Expected behavior**

Of course, end at "patching output file n/n".

**To Reproduce**

This behavior, as is common with concurrency issues, is hard to reproduce.
I had it on several projects, but not deterministically reproducible.

The following preconditions must be met:

- a pretty large project (hundreds, better thousands of C++ input files)
- set NUM_PROC_THREADS = 0
- set DOT_NUM_THREADS = 0 (this error is not in the dot stage, but I had dot errors too).
- enable a HTML build

I know it's pretty tough to find this issue without the project(s) to debug, but maybe the issue description already gives you a hint what could go wrong? Currently I'm running locally on a machine with
Intel(R) Core(TM) i7-9850H, which has 12 cores a 2 threads, and the CI job is running on a quad-core virtual CPU afaik. I can try to investigate this more if you give me enough hints.

**Version**

Problem appeared on versions:
- Doxygen 1.9.2
- Doxygen 1.9.3
- a fairly recent master (1 week old)
- all self compiled

OS: Ubuntu 18.04 and Ubuntu 20.04

**Comment by albert-github** [comment]
You wrote:
> Intel(R) Core(TM) i7-9850H, which has 12 cores a 2 threads, and the CI job is running on a quad-core virtual CPU afaik.

Does this mean that you see the problem on (both):
- a machine with Intel(R) Core(TM) i7-9850H, which has 12 cores a 2 threads, i.e. 24 threads, so doxygen will run run in multithread mode with 24 threads and the dot run part also with 24 instances of dot?
- the CI job is running on a quad-core virtual CPUm, doxygen running with 4 threads and the dot run part with 4 instances of dot?

You wrote also:
>this error is not in the dot stage, but I had dot errors too

- what is the error / warning message you got?


Small test:
- what happens when you set `DOT_NUM_THREADS=1`?

**Comment by albert-github** [comment]
This issue might be related to the review comment at https://github.com/doxygen/doxygen/pull/9024 regarding the `static int dotindex`.

**Comment by ferdymercury** [comment]
Maybe also related: https://github.com/doxygen/doxygen/issues/9155#issuecomment-1054599889

**Comment by albert-github** [comment]
@arwedus 
The issue mentioned by @ferdymercury might have the same root cause, @arwedus can you please test with the current master.

**Comment by arwedus** [comment]
@albert-github , both the dot error and the issue mentioned in the original post are gone with new doxygen master on my projects, after re-enabling DOT_NUM_THREADS        = 0 and NUM_PROC_THREADS       = 0 on a 36-core processor. Should be fine.

**Comment by albert-github** [comment]
Thanks for the feedback.

**Comment by arwedus** [comment]
Unfortunately, I still get a non-terminating doxygen. However, it's on an azure VM and I can't really debug it. What happens now is that doxygen just stops in the stage "Generating docs for...". It doesn't happen on my local machine with 36 threads, but on a rather small VM with 4 cores. All 4 cores are at close to 100% CPU usage. No clue what's happening.

**Comment by ferdymercury** [comment]
Maybe you can try in your own machine the following to see:

- Build latest doxygen in "Debug" mode
- Run "valgrind --tool=helgrind /path/to/doxygen"

Even if it terminates well in your machine, this command might give you some 'warnings' that could point at the error, and this way help for solving it in the VM.

**Comment by arwedus** [comment]
@albert-github : It's tough to do so, as the error is transient. I'm building 20+ doxygen projects, and if I do so locally, the runs complete all the time. The deadlocks just happen inside the VM and inside the CI docker image, but randomly in one of the (larger) projects.

Just wanted to let you know that the problem of infinite doxygen run time still exists - however, it seems to be a different issue as it does not occur in the "patching output file" phase any more.

I don't have gdb in the docker image, I don't build doxygen debug for the docker image, and I can't start the build with gdb all the time in CI/CD environment - also because doxygen seems to go into deadlock in different projects, not always the same. It could also be that doxygen would not go into deadlock if built in debug mode, not sure.

I tried to attach via gdb to the hanging doxygen instance and it only shows me that it's stuck in "futex_wait" (called via "pthread_cond_destroy"). Everything else is "??".

**Comment by albert-github** [comment]
Recently some problems were fixed in the master version, maybe a good idea to give it a try again.

**Comment by arwedus** [comment]
I've let our CI loose on doxygen master build `1.9.4 (482d4caeed7524c665e6ea3319fc989a8b3471f3*)`, do you mean even more recent?

**Comment by albert-github** [comment]
I cannot find the 482d4caeed7524c665e6ea3319fc989a8b3471f3 version (so I don't know when this one was created) but I would go for the latest 1.9.4 version (at the moment of writing: 0d5a456a710dde47801fba67592c6d1b33071b70)

**Comment by arwedus** [comment]
I think I meant commit 36dae95b001a87d1ecf46e2f044b9a138236dc67 - I don't see anything multitasking related changed since then, or am I in error? I could try to build debug doxygen for the CICD environment and then get a meaningful stack trace when the error appears again.

**Comment by albert-github** [comment]
It indeed doesn't look like that there are multi threading related issues, though some fixes can have side, positive, effects, so to be on the safe side...

**Comment by arwedus** [comment]
Interesting, I also end up in the same deadlock when `NUM_PROC_THREADS=1` with the current doxygen build.

**Comment by arwedus** [comment]
I overlooked an error message I get. I think it's totally unrelated to multi-thread build and I will create a new issue. Thanks for pointing me to the speedup, though, I'll give it a try, too.

**Comment by albert-github** [comment]
@arwedus Is this issue fixed nowadays as well (seen a.o. the results of #9286 and #9283) or is there still a problem here?

**Comment by arwedus** [comment]
@albert-github , I just ran multiple large projects on our CI build jobs again and it terminates, so I would mark this issue as fixed, too. Thanks!

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
